tests/common/dualtor/dual_tor_io.py

```diff
@@ -0,0 +1,424 @@
+
+        logger.info('Fetching log files from ptf and dut hosts')
+        logs_list =  [
+            {'src': sniffer_log, 'dest': '/tmp/', 'flat': True, 'fail_on_missing': False},
+            {'src': capture_pcap, 'dest': '/tmp/', 'flat': True, 'fail_on_missing': False}
+        ]
+
+        for log_item in logs_list:
+            self.ptfhost.fetch(**log_item)
+        
+        self.all_packets = scapyall.rdpcap(capture_pcap)
+        logger.info("Number of all packets captured: {}".format(len(self.all_packets)))
+
+
+    def get_total_disruptions(self):
+        return self.disrupts_count
+
+
+    def get_longest_disruption(self):
+        return self.max_disrupt_time
+
+
+    def get_total_disrupted_packets(self):
+        return self.total_disrupt_packets
+
+
+    def no_flood(self, packet):
+        """
+        This method filters packets which are unique (i.e. no floods).
+        """
+        if (not int(str(packet[scapyall.TCP].payload).replace('X','')) in self.unique_id) and (packet[scapyall.Ether].src == self.dut_mac):
+            # This is a unique (no flooded) received packet.
+            self.unique_id.append(int(str(packet[scapyall.TCP].payload).replace('X','')))
+            return True
+        elif packet[scapyall.Ether].dst == self.dut_mac:
+            # This is a sent packet.
+            return True
+        else:
+            return False
+
+
+    def examine_flow(self):
+        """
+        This method examines packets collected by sniffer thread
+        The method compares TCP payloads of the packets one by one (assuming all payloads are consecutive integers),
+        and the losses if found - are treated as disruptions in Dataplane forwarding.
+        All disruptions are saved to self.lost_packets dictionary, in format:
+        disrupt_start_id = (missing_packets_count, disrupt_time, disrupt_start_timestamp, disrupt_stop_timestamp)
+        """
+        def examine_each_packet(dut_mac, packets):
+            lost_packets = dict()
+            prev_payload, prev_time = 0, 0
+            sent_payload = 0
+            disruption_start, disruption_stop = None, None
+            received_counter = 0    # Counts packets from dut.
+            for packet in packets:
+                if packet[scapyall.Ether].dst == dut_mac:
+                    # This is a sent packet - keep track of it as payload_id:timestamp.
+                    sent_payload = int(str(packet[scapyall.TCP].payload).replace('X',''))
+                    sent_packets[sent_payload] = packet.time
+                    continue
+                if packet[scapyall.Ether].src == dut_mac:
+                    # This is a received packet.
+                    received_time = packet.time
+                    received_payload = int(str(packet[scapyall.TCP].payload).replace('X',''))
+                    received_counter += 1
+                if not (received_payload and received_time):
+                    # This is the first valid received packet.
```

Comment:
Should we verify that the first packet we receive has payload = 1?

Reply:
The first packet actually has payload 0 and timestamp as 0 - which is being checked here. May be more explicit check is better?

Reply:
Misunderstood the first time, looks good to me